contracts/interchain-token/src/contract.rs

```diff
@@ -1,32 +1,365 @@
+                },
+                |allowance| {
+                    if allowance.expiration_ledger < env.ledger().sequence() {
+                        AllowanceValue {
+                            amount: 0,
+                            expiration_ledger: allowance.expiration_ledger,
+                        }
+                    } else {
+                        allowance
+                    }
+                },
+            )
+    }
+
+    fn write_allowance(
+        env: &Env,
+        from: Address,
+        spender: Address,
+        amount: i128,
+        expiration_ledger: u32,
+    ) {
+        let allowance = AllowanceValue {
+            amount,
+            expiration_ledger,
+        };
+
+        assert_with_error!(
+            !(amount > 0 && expiration_ledger < env.ledger().sequence()),
+            "expiration_ledger is less than ledger seq when amount > 0"
+        );
+
+        let key = DataKey::Allowance(AllowanceDataKey { from, spender });
+        env.storage().temporary().set(&key, &allowance);
+
+        if amount > 0 {
+            let live_for = expiration_ledger
+                .checked_sub(env.ledger().sequence())
+                .unwrap();
+
+            env.storage()
+                .temporary()
+                .extend_ttl(&key, live_for, live_for)
+        }
+    }
+
+    fn spend_allowance(env: &Env, from: Address, spender: Address, amount: i128) {
+        let allowance = Self::read_allowance(env, from.clone(), spender.clone());
+
+        assert_with_error!(allowance.amount >= amount, "insufficient allowance");
+
+        if amount > 0 {
+            Self::write_allowance(
+                env,
+                from,
+                spender,
+                allowance
+                    .amount
+                    .checked_sub(amount)
+                    .expect("insufficient allowance"),
+                allowance.expiration_ledger,
+            );
+        }
+    }
+
+    fn read_balance(env: &Env, addr: Address) -> i128 {
+        let key = DataKey::Balance(addr);
+        env.storage()
+            .persistent()
+            .get::<DataKey, i128>(&key)
+            .map_or(0, |balance| {
+                Self::extend_balance_ttl(env, &key);
+                balance
+            })
+    }
+
+    fn receive_balance(env: &Env, addr: Address, amount: i128) {
+        let key = DataKey::Balance(addr);
+
+        env.storage()
+            .persistent()
+            .update(&key, |balance: Option<i128>| {
+                balance.unwrap_or_default() + amount
+            });
+    }
+
+    fn spend_balance(env: &Env, addr: Address, amount: i128) {
+        let balance = Self::read_balance(env, addr.clone());
+
+        assert_with_error!(balance >= amount, "insufficient balance");
+
+        Self::write_balance(env, addr, balance - amount);
+    }
+
+    fn write_metadata(env: &Env, metadata: TokenMetadata) {
+        TokenUtils::new(env).metadata().set_metadata(&metadata);
```

Comment:
what datakey does `set_metadata` use? want to ensure it doesn't collide (add a comment in DataKey with the same variant for clarity that it's in use)

Reply:
It's using `const METADATA_KEY: Symbol = symbol_short!("METADATA");` as key.

```
impl Metadata {
    pub fn new(env: &Env) -> Metadata {
        Metadata { env: env.clone() }
    }

    #[inline(always)]
    pub fn set_metadata(&self, metadata: &TokenMetadata) {
        self.env.storage().instance().set(&METADATA_KEY, metadata);
    }

    #[inline(always)]
    pub fn get_metadata(&self) -> TokenMetadata {
        self.env
            .storage()
            .instance()
            .get(&METADATA_KEY)
            .unwrap_optimized()
    }
}

```

added comment to the datastorage.